src/crypto/variant2_int_sqrt.h

```diff
@@ -0,0 +1,75 @@
+#ifndef VARIANT2_INT_SQRT_H
+#define VARIANT2_INT_SQRT_H
+
+#include <math.h>
+#include <float.h>
+
+#define VARIANT2_INTEGER_MATH_SQRT_FIXUP(r) \
+  do { \
+    const uint64_t s = r >> 1; \
+    const uint64_t b = r & 1; \
+    const uint64_t r2 = (uint64_t)(s) * (s + b) + (r << 32) - sqrt_input; \
+    r += (((int64_t)(r2 + b) > 0) ? -1 : 0) + (((int64_t)(r2 + s + (1ULL << 32)) < 0) ? 1 : 0); \
```

Comment:
Can the `(((int64_t)(r2 + s + (1ULL << 32)) < 0) ? 1 : 0)` branch really be covered? I'm implementing this in Go and found this cannot be covered while I'm using the same test code. https://codecov.io/gh/Equim-chan/cryptonight/src/8fbc190cb5d6928c1c7f68dd916261bc3892f023/v2_sqrt.go#L21

Reply:
It's casted to int64_t first then compared to 0, so yes. This code will need comments because it's not obvious how it works. I'm not sure how it should be written in Go, but it works in plain C.

Reply:
This code essentially checks these inequalities:
1) (r/2+2^32)^2 <= n+2^64
2) ((r+1)/2+2^32)^2 > n+2^64

And corrects r if needed. But it does some bit manipulations to stay within 64 bit integer types and to work both for even and odd values of r.

Reply:
The logic may make sense, but the test just doesn't cover that.

I tried this:

```diff
diff --git a/src/crypto/variant2_int_sqrt.h b/src/crypto/variant2_int_sqrt.h
index 211f0c3c..7aae4827 100644
--- a/src/crypto/variant2_int_sqrt.h
+++ b/src/crypto/variant2_int_sqrt.h
@@ -9,7 +9,7 @@
     const uint64_t s = r >> 1; \
     const uint64_t b = r & 1; \
     const uint64_t r2 = (uint64_t)(s) * (s + b) + (r << 32) - sqrt_input; \
-    r += (((int64_t)(r2 + b) > 0) ? -1 : 0) + (((int64_t)(r2 + s + (1ULL << 32)) < 0) ? 1 : 0); \
+    r += (((int64_t)(r2 + b) > 0) ? -1 : 0); \
   } while(0)
```

and it still passes all the tests, indicating that it is not covered.